// ntd/index.js
const db = require('./db')
const inquirer = require('inquirer')


const api = {
  markAsDone: async (list,index)=>{
    list[index].done = true
    await db.write(list)
  },
  markAsUndone:async (list,index)=>{
    list[index].done = false
    await db.write(list)
  },
  updateTitle:async (list,index)=>{
    inquirer.prompt({
      type: 'input',
      name: 'newTitle',
      message: "请更新你的待办事项",
    }).then(async (answer) => {
      const {newTitle} = answer
      list[index].title = newTitle
      await db.write(list)
    });
  },
  remove:async (list,index)=>{
    list.splice(index, 1)
    await db.write(list)
  },
  operateTask(list,index){
    inquirer.prompt([{
      type: 'list',
      name: 'action',
      message: '请选择一个操作',
      choices: [
        {name: '标记为已完成', value: 'markAsDone'},
        {name: '标记为未完成', value: 'markAsUndone'},
        {name: '重命名', value: 'updateTitle'},
        {name: '删除', value: 'remove'},
        '退出',
      ]
    }]).then(async (answer) => {
      const {action} = answer
      const hash = {
        markAsDone:this.markAsDone,
        markAsUndone:this.markAsUndone,
        updateTitle:this.updateTitle,
        remove:this.remove
      }
      if (hash[action]){
        hash[action](list,index)
      }
    })
  },
  createNewTask(list) {
    inquirer.prompt({
      type: 'input',
      name: 'title',
      message: "请创建你的待办事项",
    }).then(async (answer) => {
      const {title} = answer
      list.push({title: title, done: false})
      await db.write(list)
    });
  },
  showAction(list, index) {
    if (index >= 0) {
    this.operateTask(list,index)
    }

    if (index === -1) {
      this.createNewTask(list, index)
    }
  },
  showAll: async function () {
    const list = await db.read()
    inquirer
      .prompt([
        {
          type: 'list',
          name: 'index',
          message: '请选择一个待办事项',
          choices: [
            '退出',
            ...list.map((item, index) => {
              return {name: `${index + 1}. ${item.done ? "[v]" : "[_]"} ${item.title}`, value: index.toString()}
              //1. [_] 买水
            }),
            {name: '+ 添加新的待办事项', value: '-1'}
          ],
        },
      ])
      .then((answer) => {
        const index = parseInt(answer.index)
        this.showAction(list, index)
      });
  }
}
module.exports = api
